ext_authz: support for buffering request body

gsagula opened the pull request:
This PR adds support to `ext_authz` filter for buffering the request data. This is useful when the authorization server needs to check the request body, e.g. HMAC validation.

Fixes https://github.com/envoyproxy/envoy/issues/5676
 
*Risk Level*: low
*Testing*: unit
*Docs Changes*: yes
*Release Notes*: yes

htuch commented:
Please merge master to pick up https://github.com/envoyproxy/envoy/pull/5827.

lizan commented on source/extensions/filters/http/ext_authz/ext_authz.cc:
```diff
@@ -12,6 +12,20 @@ namespace Extensions {
+// Asserts that the HTTP method is for a header-only request. For details
+// https://tools.ietf.org/html/rfc7231#section-4.3.
+bool isHeaderOnlyMethod(absl::string_view method) {
+  const static std::vector<std::string>* keys = new std::vector<std::string>(
```
`std::unordered_set` or `absl::flat_hash_set` here and find for below

dio replied:
:+1:

lizan commented on api/envoy/config/filter/http/ext_authz/v2/ext_authz.proto:
```diff
@@ -38,6 +38,17 @@ message ExtAuthz {
+// Configuration for buffering the request data.
+message BufferSettings {
+  // Sets the maximum size of a message body that the filter will hold in memory. Envoy will return
+  // *HTTP 413* and will *not* initiate the authorization process when buffer reaches the number
```
It might be orthogonal though, will there be some use cases that only require first certain number of bytes, so instead of 413, sending first x bytes to authz server?

gsagula replied:
👍

gsagula replied:
Still working on this one. Missing unit tests as well as some manual verification.

lizan commented on api/envoy/service/auth/v2/attribute_context.proto:
```diff
@@ -102,6 +103,9 @@ message AttributeContext {
     string protocol = 10;
+
+    // The HTTP request body.
+    envoy.api.v2.core.DataSource body = 11;
```
not sure DataSource is appropriate here, pointing to a file doesn't make sense unless authz server are on the same box. `bytes` or `string` is enough?

gsagula replied:
👍

lizan commented on source/extensions/filters/common/ext_authz/ext_authz_http_impl.cc:
```diff
@@ -161,7 +161,17 @@ void RawHttpClientImpl::check(RequestCallbacks& callbacks,
+  const uint64_t request_length =
+      request.attributes().request().http().body().inline_bytes().length();
+  if (request_length > 0) {
+    const Http::HeaderMap& header_map =
```
combine this line with the line below, or make this non-reference and use `std::move` below to avoid copy

lizan commented on api/envoy/config/filter/http/ext_authz/v2/ext_authz.proto:
```diff
@@ -38,6 +38,17 @@ message ExtAuthz {
   bool failure_mode_allow = 2;
+
+  // Enables filter to buffer and add the request body to the authorization request.
+  BufferSettings with_request_body = 4;
```
Add a note on how this play with `failure_mode_allow`? seems with current impl, if request body exceeds the buffer setting it will result 413 regardless of `failure_mode_allow`.

gsagula commented:
@lizan Thank you for looking at the PR. I had to work on some other things, but I will work on this by next week.

gsagula commented:
Sorry, another busy week. I will take a look at this later today or Monday.

gsagula commented:
@lizan I've addressed all your comments.

gsagula commented on source/extensions/filters/common/ext_authz/check_request_utils.cc:
```diff
@@ -114,20 +114,29 @@ void CheckRequestUtils::setHttpRequest(
+  const Buffer::Instance* buffer = sdfc->decodingBuffer();
+  if (max_request_bytes > 0 && buffer != nullptr) {
+    const uint64_t len = std::min(buffer->length(), max_request_bytes);
+    std::unique_ptr<char[]> data(new char[len]);
```
I think I could use an `std::string` and then just move it. Not sure.

lizan replied:
yeah I think you can do that.

lizan commented on source/extensions/filters/common/ext_authz/ext_authz_http_impl.cc:
```diff
@@ -161,7 +161,17 @@ void RawHttpClientImpl::check(RequestCallbacks& callbacks,
   callbacks_ = &callbacks;
 
-  Http::HeaderMapPtr headers = std::make_unique<Http::HeaderMapImpl>(lengthZeroHeader());
+  Http::HeaderMapPtr headers{};
```
nit: no need `{}`.

gsagula replied:
👍

lizan commented on source/extensions/filters/http/ext_authz/ext_authz.cc:
```diff
@@ -7,11 +7,27 @@
+       Http::Headers::get().MethodValues.Connect, Http::Headers::get().MethodValues.Options,
+       Http::Headers::get().MethodValues.Trace});
+
+  return std::any_of(keys->begin(), keys->end(),
```
`keys->find(method) != keys->end()`

gsagula replied:
👍

lizan commented on source/extensions/filters/http/ext_authz/ext_authz.cc:
```diff
@@ -65,14 +82,36 @@ void Filter::initiateCall(const Http::HeaderMap& headers) {
+
+    // Call the authorization service.
+    ENVOY_STREAM_LOG(debug, "ext_authz finished buffering", *callbacks_);
+    initiateCall(*request_headers_);
```
If the request is small and with trailer (i.e. end_stream is false), then `initiateCall` won't be called? Add a test case for that.

gsagula replied:
👍 Great question! Thanks

gsagula commented:
I will try to get back to this one tomorrow. Thanks for reviewing it.

lizan commented on source/extensions/filters/common/ext_authz/check_request_utils.cc:
```diff
@@ -118,13 +119,24 @@ void CheckRequestUtils::setHttpRequest(
     std::string data;
-    data.reserve(len);
-    buffer->copyOut(0, len, &data[0]);
+    data.reserve(length);
```
Instead of `reserve` if you do `resize` you should be able to do copyOut to &data[0].

gsagula replied:
Interesting! I was hitting a segfault. I think I could also initialize the string with the correct length, so it doesn't reallocate memory.

dio commented on docs/root/intro/version_history.rst:
```diff
@@ -20,6 +20,9 @@ Version history
 * cors: added :ref:`filter_enabled & shadow_enabled RuntimeFractionalPercent flags <cors-runtime>` to filter.
+* ext_authz: added support for buffering request body.
+* ext_authz: migrated from V2alpha to V2 and improved docs.
+* ext_authz: added an configurable option to make the gRPC service cross-compatible with V2Alpha. Note that this feature is already deprecated. It should be used for a short time, and only when transitioning from alpha to V2 release version. 
```
s/added an configurable/added a configurable

dio commented on docs/root/intro/version_history.rst:
```diff
@@ -20,6 +20,9 @@ Version history
 * config: added support for :ref:`initial_fetch_timeout <envoy_api_field_core.ConfigSource.initial_fetch_timeout>`. The timeout is disabled by default.
 * cors: added :ref:`filter_enabled & shadow_enabled RuntimeFractionalPercent flags <cors-runtime>` to filter.
+* ext_authz: added support for buffering request body.
+* ext_authz: migrated from V2alpha to V2 and improved docs.
```
Should we use lowercase instead? `v2alpha`, `v2`, and the same elsewhere.

Just for consistency.

dio commented on source/extensions/filters/common/ext_authz/ext_authz_http_impl.cc:
```diff
@@ -161,7 +161,17 @@ void RawHttpClientImpl::check(RequestCallbacks& callbacks,
 
-  Http::HeaderMapPtr headers = std::make_unique<Http::HeaderMapImpl>(lengthZeroHeader());
+  Http::HeaderMapPtr headers;
+  const uint64_t request_length = request.attributes().request().http().body().size();
```
😄

gsagula commented:
@lizan @dio I think I addressed all your comments. Thank you for reviewing it.

brectanus-sigsci commented:
@gsagula Any update/ETA on this? Seems all comments were addressed.

itstehkman commented:
Same! I would love for this to be merged, am also waiting on it as it will help me a lot, trying to do HMAC in ext-authz :)

lizan commented on source/extensions/filters/http/ext_authz/ext_authz.cc:
```diff
@@ -7,11 +7,26 @@
+// Asserts that the HTTP method is for a header-only request. For details
+// https://tools.ietf.org/html/rfc7231#section-4.3.
+bool isHeaderOnlyMethod(absl::string_view method) {
+  const static absl::flat_hash_set<std::string>* keys = new absl::flat_hash_set<std::string>(
```
I roughly remember GET doesn’t have to be header only and I couldn’t find out where this is coning from in the RFC, more comments?

gsagula replied:
No, you are right, but I think we will have problems when the stream is doing a WebSocket upgrade. I will need to look into it.
```
A payload within a GET request message has no defined semantics;
sending a payload body on a GET request might cause some existing
implementations to reject the request.
```

gsagula commented:
@lizan No worries, I've been pretty busy lately too. I found a couple of issues with the partial message feature, but I need to fix the broken tests now. I will try to get it out tomorrow. Thanks!

gsagula commented:
Build failures seem unrelated. I will merge master and retry.

lizan commented on source/extensions/filters/http/ext_authz/ext_authz.cc:
```diff
@@ -72,23 +58,22 @@ void Filter::initiateCall(const Http::HeaderMap& headers) {
   buffer_data_ = config_->withRequestBody() &&
-                 !(end_stream || (method && isHeaderOnlyMethod(method->value().getStringView())));
-
+                 !(end_stream || Http::Utility::isWebSocketUpgradeRequest(headers) ||
```
Do filters get upgrade requests? I roughly remember this is handled by HCM now? cc @alyssawilk

gsagula replied:
Apparently, they do but would be nice to hear from @alyssawilk. If this is ok, the same logic ^^ should be applied to the buffer filter.

lizan replied:
Ah nvm I was in the old impression before we have new style websocket.

gsagula commented:
@lizan @dio I have addressed your concerns I believe. I also added a few tests for the logic around upgrade requests and did some manual testing. Feel free to take another look. Thanks!
